frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart

```diff
@@ -0,0 +1,279 @@
+
+  @override
+  void initState() {
+    super.initState();
+
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      _focusNode.requestFocus();
+      _searchEmoji();
+    });
+  }
+
+  @override
+  void dispose() {
+    _emojiController.dispose();
+    _focusNode.dispose();
+    super.dispose();
+  }
+
+  KeyEventResult _onKey(FocusNode node, RawKeyEvent event) {
+    // Check if the key event is key press, not key release
+    if (event is! RawKeyDownEvent) return KeyEventResult.ignored;
+
+    // Handle arrow keys
+    if (arrowKeys[event.logicalKey] != null) {
+      // Computing new emoji selection index
+      final int newSelectedIndex =
+          (_selectedIndex + arrowKeys[event.logicalKey]!)
+              .clamp(0, searchEmojiList.length - 1);
+
+      if (newSelectedIndex == _selectedIndex) return KeyEventResult.ignored;
+
+      setState(() => _selectedIndex = newSelectedIndex);
+      return KeyEventResult.handled;
+    }
+
+    switch (event.logicalKey) {
+      case LogicalKeyboardKey.enter:
+        _selectEmoji();
+        return KeyEventResult.handled;
+
+      case LogicalKeyboardKey.backspace:
+        // Delete a character on the editor
+        widget.editorState.deleteBackward();
+        if (_emojiController.text.isEmpty) {
+          widget.onExit();
+          return KeyEventResult.handled;
+        }
+
+        // Delete a character on the editor
+        _emojiController.text = _emojiController.text
+            .substring(0, _emojiController.text.length - 1);
+        _searchEmoji();
+        return KeyEventResult.handled;
+
+      case LogicalKeyboardKey.escape || LogicalKeyboardKey.tab:
+        widget.onExit();
+        return KeyEventResult.handled;
+
+      default:
+
+        // Determine whether or not the event was from a key with a character
+        if (event.character == null) return KeyEventResult.ignored;
+
+        // Typing character
+        widget.editorState.insertTextAtCurrentSelection(event.character!);
+
+        // Determine whether or not the shortcut character was pressed again
+        if (event.character == widget.shortcutCharacter) {
+          widget.onExit();
+          return KeyEventResult.handled;
+        }
+
+        // Add the character
+        _emojiController.text += event.character!;
+        _searchEmoji();
+        return KeyEventResult.handled;
+    }
+  }
+
+  void _searchEmoji() {
+    // Formatting the emoji search text to increase chances of finding a match
+    final String query = _emojiController.text
+      ..toLowerCase()
+      ..replaceAll(" ", "_");
+
+    searchEmojiList.clear();
+
+    // Search for the emoji in the emoji category groups
+    for (int index = 0; index < widget.state.emojiCategoryGroupList.length;) {
+      searchEmojiList.addAll(
+        widget.state.emojiCategoryGroupList[index].emoji
+            .where(
+              (item) =>
+                  (query.isEmpty || item.name.toLowerCase().contains(query)) &&
```

Comment:
@Xazin, Do you think the `_searchEmoji` function is okay to approve? Finding emojis without using `emoji.name.toLowerCase().contains(query)` could be improved. However, we use something similar in the `DeafultEmojiPicker` widget. Do you think I should optimize it?